bee-network/bee-autopeering/src/server.rs

```diff
@@ -0,0 +1,290 @@
+            bind_addr: config.bind_addr(),
+        }
+    }
+}
+
+pub(crate) struct IncomingPacketSenders {
+    pub(crate) discovery_tx: IncomingPacketTx,
+    pub(crate) peering_tx: IncomingPacketTx,
+}
+
+pub(crate) struct Server {
+    config: ServerConfig,
+    local: Local,
+    incoming_senders: IncomingPacketSenders,
+    outgoing_rx: OutgoingPacketRx,
+}
+
+impl Server {
+    pub fn new(config: ServerConfig, local: Local, incoming_senders: IncomingPacketSenders) -> (Self, ServerTx) {
+        let (outgoing_tx, outgoing_rx) = server_chan::<OutgoingPacket>();
+
+        (
+            Self {
+                config,
+                local,
+                incoming_senders,
+                outgoing_rx,
+            },
+            outgoing_tx,
+        )
+    }
+
+    pub async fn init<S: PeerStore, const N: usize>(self, task_mngr: &mut TaskManager<S, N>) {
+        let Server {
+            config,
+            local,
+            incoming_senders,
+            outgoing_rx,
+        } = self;
+
+        // Try to bind the UDP socket to the configured address.
+        let socket = UdpSocket::bind(&config.bind_addr)
+            .await
+            .expect("error binding udp socket");
+
+        // The Tokio docs explain that there's no split method, and that we have to arc the UdpSocket in order to share
+        // it.
+        let incoming_socket = Arc::new(socket);
+        let outgoing_socket = Arc::clone(&incoming_socket);
+
+        let incoming_packet_handler = IncomingPacketHandler {
+            incoming_socket,
+            incoming_senders,
+            bind_addr: config.bind_addr,
+        };
+
+        let outgoing_packet_handler = OutgoingPacketHandler {
+            outgoing_socket,
+            outgoing_rx,
+            local,
+            bind_addr: config.bind_addr,
+        };
+
+        task_mngr.run::<IncomingPacketHandler>(incoming_packet_handler);
+        task_mngr.run::<OutgoingPacketHandler>(outgoing_packet_handler);
+    }
+}
+
+struct IncomingPacketHandler {
+    incoming_socket: Arc<UdpSocket>,
+    incoming_senders: IncomingPacketSenders,
+    bind_addr: SocketAddr,
+}
+
+struct OutgoingPacketHandler {
+    outgoing_socket: Arc<UdpSocket>,
+    outgoing_rx: OutgoingPacketRx,
+    local: Local,
+    bind_addr: SocketAddr,
+}
+
+#[async_trait::async_trait]
+impl Runnable for IncomingPacketHandler {
+    const NAME: &'static str = "IncomingPacketHandler";
+    const SHUTDOWN_PRIORITY: u8 = 2;
+
+    type ShutdownSignal = ShutdownRx;
+
+    async fn run(self, mut shutdown_rx: Self::ShutdownSignal) {
+        let IncomingPacketHandler {
+            incoming_socket,
+            incoming_senders,
+            bind_addr,
+        } = self;
+
+        let mut packet_bytes = [0; READ_BUFFER_SIZE];
+
+        let IncomingPacketSenders {
+            discovery_tx,
+            peering_tx,
+        } = incoming_senders;
+
+        'recv: loop {
+            tokio::select! {
+                _ = &mut shutdown_rx => {
```

Comment:
This feels like something that could be done using `ShutdownStream`.

Reply:
Yes, I'll switch to `ShutdownStream` in a future release. Want to keep it bare bones for now.